dbms/src/Flash/Mpp/MPPTunnel.h

```diff
@@ -233,23 +241,96 @@ class AsyncTunnelSender : public TunnelSender
     GRPCSendQueue<TrackedMppDataPacketPtr> queue;
 };
 
-/// LocalTunnelSender just provide readForLocal method to return one element one time
-/// LocalTunnelSender is owned by the associated ExchangeReceiver
+template <bool enable_fine_grained_shuffle>
 class LocalTunnelSender : public TunnelSender
 {
 public:
-    using Base = TunnelSender;
-    using Base::Base;
-    TrackedMppDataPacketPtr readForLocal();
+    LocalTunnelSender(
+        size_t source_index_,
+        const String & req_info_,
+        ExchangeReceiverBase * recv_base_,
+        const LoggerPtr & log_,
+        size_t queue_size,
+        MemoryTrackerPtr & memory_tracker_,
+        const String & tunnel_id_)
+        : TunnelSender(queue_size, memory_tracker_, log_, tunnel_id_, nullptr)
+        , source_index(source_index_)
+        , req_info(req_info_)
+        , recv_base(recv_base_)
+        , channel_writer(&(recv_base->getMsgChannels()), req_info, log_, recv_base_->getDataSizeInQueue())
+        , is_done(false)
+    {}
+
+    ~LocalTunnelSender() override
+    {
+        closeLocalTunnel(false, "");
+    }
+
+    bool push(TrackedMppDataPacketPtr && data) override
+    {
+        if (unlikely(checkPacketErr(data)))
+            return false;
+
+        // receiver_mem_tracker pointer will always be valid because ExchangeReceiverBase won't be destructed
+        // before all local tunnels are destructed so that the MPPTask which contains ExchangeReceiverBase and
+        // is responsible for deleting receiver_mem_tracker must be destroyed after these local tunnels.
+        MemoryTracker * receiver_mem_tracker = recv_base->getMemoryTracker();
```

Comment:
It looks like we can remove the reference to the ExchangeReceiver in the LocalTunnel.
- hold the `receiver_mem_tracker` directly
- hold a lambda like `triggered_after_done`

```
struct LocalTunnel
{
MemoryTracker * receiver_mem_tracker;
std::function<void()> triggered_after_done;
}
```

Reply:
After that, we don't have to delete the unit test of `LocalTunnel`.